src/open_dread_rando/door_locks/custom_door_types.py

```diff
@@ -375,34 +374,29 @@ def patch_material_data(self, editor: PatcherEditor, version: str):
             for mat_dat in self.data.alternate_mats:
                 create_custom_material(editor, mat_dat)
 
-    def add_weakness(self, weakness: str, new_template: dict):
-        life_funcs: list = new_template["property"]["components"]["LIFE"]["functions"]
+    def add_weakness(self, weakness: str, new_template: Bmsad) -> None:
+        life = new_template.components["LIFE"]
 
-        new_func = {
-            "name": "AddDamageSource",
-            "unk": 1,
-            "params": {
-                "Param1": {
-                    "type": "s",
-                    "value": weakness
-                }
-            }
-        }
+        new_func = ActorDefFunc.new("AddDamageSource")
+        new_func.set_param(1, weakness)
 
+        life_funcs = list(life.functions)
         life_funcs.append(new_func)
+        life.functions = life_funcs
 
-    def change_collision(self, new_template: dict):
+    def change_collision(self, new_template: Bmsad) -> None:
         if self.data.collision is None:
             return
 
-        coll_comp = new_template["property"]["components"]["COLLISION"]
-        coll_comp["dependencies"]["file"] = self.data.collision
+        coll_comp = new_template.components["COLLISION"]
+        assert isinstance(coll_comp.dependencies, dict)
+        coll_comp.dependencies["file"] = self.data.collision
 
-    def patch(self, editor: PatcherEditor, version: str = "DEFAULT"):
+    def patch(self, editor: PatcherEditor, version: str = "DEFAULT") -> None:
         template_bmsad = _template_read_shield(self.data.type.value)
```

Comment:
change this to `actordef = editor.get_parsed_file(self.data.type.value, Bmsad)` and add the following below:

```
        # remove all damage sources but hyper beam
        life_funcs = new_actordef.components["LIFE"].functions
        life_funcs = [func for func in life_funcs
                      if func.name != "AddDamageSource" or func.get_param(1) == "HYPER_BEAM"]```